### PR TITLE
Expand vkGetQueryPoolResults/vkCmdCopyQueryPoolResults validation

### LINKED ISSUE #2029: no error/warning for invalid stride in GetQueryPoolResults
According to spec:

`If more than one query is retrieved and stride is not at least as large as the size of the array of values corresponding to a single query, the values written to memory are undefined.`

This seems like something that should have some sort of message. For example:
* a pipeline statistics query has completed with 2 bits set in `pipelineStatistics`
* the query pool has 2 result sets available
* `VK_QUERY_RESULT_64_BIT` is set
* GetQueryPoolResults is called with stride = `sizeof(uint64_t)`

In this case, I'd expect at least a warning since each query will return a 2x64 array, meaning the stride must be `2 * sizeof(uint64_t)`, but there is no message.

### PR DESCRIPTION
**Pull request by mark-lunarg**
~~Added WIP tag, waiting to hear back about [gitlab issue#2380](https://gitlab.khronos.org/vulkan/vulkan/-/issues/2380).~~
Issue is approved, a stride test will be added in a follow-on PR when the new VU appears in the header.

Added some stride/datasize checks to `GetQueryPoolResults`, a check or to to `CmdQueryPoolResults`, fixed a bug or two and removed some dead code.  Added tests.  

Fixes #2029.

[Passed internal CI](http://erusea:4040/#/c/7306/).

### CODE REVIEW

**Comment by mark-lunarg** [comment]
@djdeath, if you can spare the time, would you mind taking a look and sharing your feedback?  Thanks!

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -3780,36 +3789,70 @@ bool CoreChecks::ValidateGetQueryPoolResultsFlags(VkQueryPool queryPool, VkQuery
...
+
+                case VK_QUERY_TYPE_PERFORMANCE_QUERY_KHR:
+                    // Performance queries store results in a tightly packed array whose type is determined by the unit member of
+                    // the corresponding VkPerformanceCounterKHR
+                    query_items = 0;
+                    for (uint32_t i = 0; i < query_pool_state->perf_counter_indices.size(); i++) {
+                        query_items +=
+                            (query_pool_state->perf_counter_storage_64_bit[query_pool_state->perf_counter_indices[i]] ? 2 : 1);
+                    }
+                    query_size = query_size_in_bytes * query_items;
```
Comment (Tony-LunarG) [inline layers/core_validation.cpp:3837]: The spec says "Performance queries write one VkPerformanceCounterResultKHR value for each VkPerformanceCounterKHR in the query.", so shouldn't query_size be a multiple of the sizeof(VkPerformanceCounterResultKHR)?

**Review by Tony-LunarG — COMMENTED** [review]
Question about performance query result sizes

File: layers/core_validation.cpp (hunk 1 above)
Comment (mark-lunarg) [reply layers/core_validation.cpp:3837]: Nah, you gotta put the `QueryPoolPerfCreateInfoKHR `in the `pNext `of your `QueryPoolCreateInfo `-- that tells you _which_ of the `VkPerformanceCounterResultKHRs `are referenced by `counterIndexCount/pCounterIndices`.  The query will "store results in a tightly packed array whose _type_ is determined by the unit member of the corresponding `VkPerformanceCounterKHR`".  These `VkPerformanceCounterKHR `structures are static and are returned by a separate API, `EnumeratePhysicalDeviceQueueFamilyPerformanceQueryCountersKHR`.  The 'types' are one of the `VkPerformanceCounterStorageKHR `enums, which are all 32 or 64 bits.

Comment (djdeath) [reply layers/core_validation.cpp:3837]: Our implementation writes an array of `VkPerformanceCounterResultKHR`.

The quote you pasted seems to be at odds with : 
```
Performance queries write one VkPerformanceCounterResultKHR value for each VkPerformanceCounterKHR in the query.
```

**Comment by djdeath** [comment]
Running into a crash on the test : 

```
[ RUN      ] VkLayerTest.QueryPerformanceCreation
[New Thread 0x7ffff2c7b700 (LWP 433060)]
[New Thread 0x7ffff247a700 (LWP 433061)]
[New Thread 0x7ffff1c79700 (LWP 433062)]
[New Thread 0x7ffff1478700 (LWP 433063)]
--Type <RET> for more, q to quit, c to continue without paging--

bt
Thread 1 "vk_layer_valida" received signal SIGSEGV, Segmentation fault.
std::_Bit_reference::operator= (this=0x7fffffffcb90, __x=false) at /usr/include/c++/10/bits/stl_bvector.h:95
95		*_M_p &= ~_M_mask;
(gdb) bt
#0  std::_Bit_reference::operator= (this=0x7fffffffcb90, __x=false) at /usr/include/c++/10/bits/stl_bvector.h:95
#1  0x00007ffff3c101c4 in ValidationStateTracker::PostCallRecordCreateQueryPool (this=0x5555585ae480, 
    device=0x555557e402d0, pCreateInfo=0x7fffffffcea0, pAllocator=0x0, pQueryPool=0x7fffffffce98, result=VK_SUCCESS)
    at Vulkan-ValidationLayers/layers/state_tracker.cpp:2796
#2  0x00007ffff3a726ad in vulkan_layer_chassis::CreateQueryPool (device=0x555557e402d0, pCreateInfo=0x7fffffffcea0, 
    pAllocator=0x0, pQueryPool=0x7fffffffce98)
    at Vulkan-ValidationLayers/layers/generated/chassis.cpp:1789
#3  0x00007ffff73e87ef in vkCreateQueryPool (device=0x555557e402d0, pCreateInfo=0x7fffffffcea0, pAllocator=0x0, 
    pQueryPool=0x7fffffffce98) at Vulkan-Loader/loader/trampoline.c:1130
#4  0x0000555555f0ff23 in VkLayerTest_QueryPerformanceCreation_Test::TestBody (this=0x5555580b6dd0)
    at Vulkan-ValidationLayers/tests/vklayertests_others.cpp:7432
#5  0x00007ffff7dda37a in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (
    object=0x5555580b6dd0, method=&virtual testing::Test::TestBody(), location=0x7ffff7dece63 "the test body")
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2433
#6  0x00007ffff7dd3989 in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (
    object=0x5555580b6dd0, method=&virtual testing::Test::TestBody(), location=0x7ffff7dece63 "the test body")
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2469
#7  0x00007ffff7db1b18 in testing::Test::Run (this=0x5555580b6dd0)
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2508
#8  0x00007ffff7db23e1 in testing::TestInfo::Run (this=0x555557d60b80)
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2684
#9  0x00007ffff7db2aa5 in testing::TestSuite::Run (this=0x555557d323f0)
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2816
#10 0x00007ffff7dbe017 in testing::internal::UnitTestImpl::RunAllTests (this=0x555557d2c040)
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:5338
#11 0x00007ffff7ddb545 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x555557d2c040, 
    method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x7ffff7dbdc40 <testing::internal::UnitTestImpl::RunAllTests()>, 
    location=0x7ffff7ded8a0 "auxiliary test code (environments or event listeners)")
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2433
#12 0x00007ffff7dd49e5 in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x555557d2c040, 
    method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x7ffff7dbdc40 <testing::internal::UnitTestImpl::RunAllTests()>, 
    location=0x7ffff7ded8a0 "auxiliary test code (environments or event listeners)")
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2469
#13 0x00007ffff7dbcaa2 in testing::UnitTest::Run (this=0x7ffff7e15d00 <testing::UnitTest::GetInstance()::instance>)
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:4925
#14 0x0000555555d17132 in RUN_ALL_TESTS ()
    at Vulkan-ValidationLayers/external/googletest/googletest/include/gtest/gtest.h:2473
#15 0x0000555555d12c85 in main (argc=1, argv=0x7fffffffd6e8)
    at Vulkan-ValidationLayers/tests/layer_validation_tests.cpp:3379
(gdb) 
#0  std::_Bit_reference::operator= (this=0x7fffffffcb90, __x=false) at /usr/include/c++/10/bits/stl_bvector.h:95
#1  0x00007ffff3c101c4 in ValidationStateTracker::PostCallRecordCreateQueryPool (this=0x5555585ae480, 
    device=0x555557e402d0, pCreateInfo=0x7fffffffcea0, pAllocator=0x0, pQueryPool=0x7fffffffce98, result=VK_SUCCESS)
    at Vulkan-ValidationLayers/layers/state_tracker.cpp:2796
#2  0x00007ffff3a726ad in vulkan_layer_chassis::CreateQueryPool (device=0x555557e402d0, pCreateInfo=0x7fffffffcea0, 
    pAllocator=0x0, pQueryPool=0x7fffffffce98)
    at Vulkan-ValidationLayers/layers/generated/chassis.cpp:1789
#3  0x00007ffff73e87ef in vkCreateQueryPool (device=0x555557e402d0, pCreateInfo=0x7fffffffcea0, pAllocator=0x0, 
    pQueryPool=0x7fffffffce98) at Vulkan-Loader/loader/trampoline.c:1130
#4  0x0000555555f0ff23 in VkLayerTest_QueryPerformanceCreation_Test::TestBody (this=0x5555580b6dd0)
    at Vulkan-ValidationLayers/tests/vklayertests_others.cpp:7432
#5  0x00007ffff7dda37a in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test, void> (
    object=0x5555580b6dd0, method=&virtual testing::Test::TestBody(), location=0x7ffff7dece63 "the test body")
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2433
#6  0x00007ffff7dd3989 in testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void> (
    object=0x5555580b6dd0, method=&virtual testing::Test::TestBody(), location=0x7ffff7dece63 "the test body")
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2469
#7  0x00007ffff7db1b18 in testing::Test::Run (this=0x5555580b6dd0)
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2508
#8  0x00007ffff7db23e1 in testing::TestInfo::Run (this=0x555557d60b80)
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2684
#9  0x00007ffff7db2aa5 in testing::TestSuite::Run (this=0x555557d323f0)
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2816
#10 0x00007ffff7dbe017 in testing::internal::UnitTestImpl::RunAllTests (this=0x555557d2c040)
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:5338
#11 0x00007ffff7ddb545 in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x555557d2c040, 
    method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x7ffff7dbdc40 <testing::internal::UnitTestImpl::RunAllTests()>, 
    location=0x7ffff7ded8a0 "auxiliary test code (environments or event listeners)")
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2433
#12 0x00007ffff7dd49e5 in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (object=0x555557d2c040, 
    method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0x7ffff7dbdc40 <testing::internal::UnitTestImpl::RunAllTests()>, 
    location=0x7ffff7ded8a0 "auxiliary test code (environments or event listeners)")
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:2469
#13 0x00007ffff7dbcaa2 in testing::UnitTest::Run (this=0x7ffff7e15d00 <testing::UnitTest::GetInstance()::instance>)
    at Vulkan-ValidationLayers/external/googletest/googletest/src/gtest.cc:4925
#14 0x0000555555d17132 in RUN_ALL_TESTS ()
    at Vulkan-ValidationLayers/external/googletest/googletest/include/gtest/gtest.h:2473
#15 0x0000555555d12c85 in main (argc=1, argv=0x7fffffffd6e8)
    at Vulkan-ValidationLayers/tests/layer_validation_tests.cpp:3379

```

File: layers/core_validation.cpp (hunk 1 above)
Comment (mark-lunarg) [reply layers/core_validation.cpp:3837]: @Tony-LunarG, @djdeath, you guys are absolutely right -- I confused the structures. Code reviews are good!

Comment (mark-lunarg) [reply layers/core_validation.cpp]: OK, new version up, please have another look.

**Comment by djdeath** [comment]
```
$ vkbuild_run ./build/tests/vk_layer_validation_tests --gtest_filter=VkLayerTest.QueryPerf*
Note: Google Test filter = VkLayerTest.QueryPerf*
[==========] Running 6 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 6 tests from VkLayerTest
[ RUN      ] VkLayerTest.QueryPerformanceCreation
[       OK ] VkLayerTest.QueryPerformanceCreation (31 ms)
[ RUN      ] VkLayerTest.QueryPerformanceCounterCommandbufferScope
             TEST SKIPPED: No queue reported any performance counter with command buffer scope.
[       OK ] VkLayerTest.QueryPerformanceCounterCommandbufferScope (26 ms)
[ RUN      ] VkLayerTest.QueryPerformanceCounterRenderPassScope
             TEST SKIPPED: No queue reported any performance counter with render pass scope.
[       OK ] VkLayerTest.QueryPerformanceCounterRenderPassScope (26 ms)
[ RUN      ] VkLayerTest.QueryPerformanceReleaseProfileLockBeforeSubmit
[       OK ] VkLayerTest.QueryPerformanceReleaseProfileLockBeforeSubmit (36 ms)
[ RUN      ] VkLayerTest.QueryPerformanceIncompletePasses
/home/djdeath/src/mesa-src/Vulkan-ValidationLayers/tests/vkrenderframework.cpp:208: Failure
Failed
Expected to succeed but got error: Validation Error: [ VUID-vkGetQueryPoolResults-dataSize-00817 ] Object 0: handle = 0x90000000009, type = VK_OBJECT_TYPE_QUERY_POOL; | MessageID = 0x73375ed4 | vkGetQueryPoolResults() on querypool VkQueryPool 0x90000000009[] specified dataSize as 0, which is too small for the specified query type and options. The Vulkan spec states: dataSize must be large enough to contain the result of each query, as described here (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkGetQueryPoolResults-dataSize-00817)
[  FAILED  ] VkLayerTest.QueryPerformanceIncompletePasses (49 ms)
[ RUN      ] VkLayerTest.QueryPerformanceResetAndBegin
[       OK ] VkLayerTest.QueryPerformanceResetAndBegin (30 ms)
[----------] 6 tests from VkLayerTest (198 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 1 test suite ran. (198 ms total)
[  PASSED  ] 5 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] VkLayerTest.QueryPerformanceIncompletePasses

 1 FAILED TEST

```

**Comment by djdeath** [comment]
I tweaked my driver to run the skipped tests: 

```
$ vkbuild_run ./build/tests/vk_layer_validation_tests --gtest_filter=VkLayerTest.QueryPerf*
Note: Google Test filter = VkLayerTest.QueryPerf*
[==========] Running 6 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 6 tests from VkLayerTest
[ RUN      ] VkLayerTest.QueryPerformanceCreation
[       OK ] VkLayerTest.QueryPerformanceCreation (62 ms)
[ RUN      ] VkLayerTest.QueryPerformanceCounterCommandbufferScope
[       OK ] VkLayerTest.QueryPerformanceCounterCommandbufferScope (34 ms)
[ RUN      ] VkLayerTest.QueryPerformanceCounterRenderPassScope
[       OK ] VkLayerTest.QueryPerformanceCounterRenderPassScope (34 ms)
[ RUN      ] VkLayerTest.QueryPerformanceReleaseProfileLockBeforeSubmit
[       OK ] VkLayerTest.QueryPerformanceReleaseProfileLockBeforeSubmit (45 ms)
[ RUN      ] VkLayerTest.QueryPerformanceIncompletePasses
Unexpected: Validation Error: [ VUID-vkCmdEndQuery-queryPool-03227 ] Object 0: handle = 0x55773f21c480, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x2d3509b3 | vkCmdEndQuery: Query pool VkQueryPool 0x90000000009[] was created with a counter of scopeVK_QUERY_SCOPE_COMMAND_BUFFER_KHR but the end of the query is not the last command in the command buffer VkCommandBuffer 0x55773f21c480[]. The Vulkan spec states: If queryPool was created with a queryType of VK_QUERY_TYPE_PERFORMANCE_QUERY_KHR and one or more of the counters used to create queryPool was VK_PERFORMANCE_COUNTER_SCOPE_COMMAND_BUFFER_KHR, the vkCmdEndQuery must be the last recorded command in commandBuffer (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdEndQuery-queryPool-03227)
/home/djdeath/src/mesa-src/Vulkan-ValidationLayers/tests/vkrenderframework.cpp:195: Failure
Failed
Received unexpected error(s).
/home/djdeath/src/mesa-src/Vulkan-ValidationLayers/tests/vkrenderframework.cpp:208: Failure
Failed
Expected to succeed but got error: Validation Error: [ VUID-vkGetQueryPoolResults-dataSize-00817 ] Object 0: handle = 0x90000000009, type = VK_OBJECT_TYPE_QUERY_POOL; | MessageID = 0x73375ed4 | vkGetQueryPoolResults() on querypool VkQueryPool 0x90000000009[] specified dataSize as 0, which is too small for the specified query type and options. The Vulkan spec states: dataSize must be large enough to contain the result of each query, as described here (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkGetQueryPoolResults-dataSize-00817)
[  FAILED  ] VkLayerTest.QueryPerformanceIncompletePasses (52 ms)
[ RUN      ] VkLayerTest.QueryPerformanceResetAndBegin
Unexpected: Validation Error: [ VUID-vkCmdBeginQuery-queryPool-03224 ] Object 0: handle = 0x55773f143230, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x92fda600 | vkCmdBeginQuery(): Query pool VkQueryPool 0x90000000009[] was created with a counter of scope VK_QUERY_SCOPE_COMMAND_BUFFER_KHR but vkCmdBeginQuery() is not the first recorded command in the command buffer. The Vulkan spec states: If queryPool was created with a queryType of VK_QUERY_TYPE_PERFORMANCE_QUERY_KHR and one of the counters used to create queryPool was VK_PERFORMANCE_COUNTER_SCOPE_COMMAND_BUFFER_KHR, the query begin must be the first recorded command in commandBuffer (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdBeginQuery-queryPool-03224)
Unexpected: Validation Error: [ VUID-vkCmdEndQuery-queryPool-03227 ] Object 0: handle = 0x55773f143230, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x2d3509b3 | vkCmdEndQuery: Query pool VkQueryPool 0x90000000009[] was created with a counter of scopeVK_QUERY_SCOPE_COMMAND_BUFFER_KHR but the end of the query is not the last command in the command buffer VkCommandBuffer 0x55773f143230[]. The Vulkan spec states: If queryPool was created with a queryType of VK_QUERY_TYPE_PERFORMANCE_QUERY_KHR and one or more of the counters used to create queryPool was VK_PERFORMANCE_COUNTER_SCOPE_COMMAND_BUFFER_KHR, the vkCmdEndQuery must be the last recorded command in commandBuffer (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkCmdEndQuery-queryPool-03227)
/home/djdeath/src/mesa-src/Vulkan-ValidationLayers/tests/vkrenderframework.cpp:195: Failure
Failed
Received unexpected error(s).
[  FAILED  ] VkLayerTest.QueryPerformanceResetAndBegin (37 ms)
[----------] 6 tests from VkLayerTest (265 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 1 test suite ran. (266 ms total)
[  PASSED  ] 4 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] VkLayerTest.QueryPerformanceIncompletePasses
[  FAILED  ] VkLayerTest.QueryPerformanceResetAndBegin

 2 FAILED TESTS
```

Seems to trigger additional issues.

**Comment by mark-lunarg** [comment]
@djdeath, what exactly was the difference between the two above sets of test output?  For instance, `QueryPerformanceResetAndBegin `passed in the first set and failed in the second.

**Comment by djdeath** [comment]
> @djdeath, what exactly was the difference between the two above sets of test output? For instance, `QueryPerformanceResetAndBegin `passed in the first set and failed in the second.

All of our counters are command scoped, I manually switch a couple to be render pass & command buffer scoped.

**Comment by mark-lunarg** [comment]
@djdeath, pushed a potential fix for the 817 unexpected error.  I believe the other issues may be unrelated to this change, and we'd probably see them on Master as well.  I don't have access to PerfQuery-capable HW but the source for the `QueryPerformanceIncompletePasses `test seems to have `EndQuery `as the last command in `m_commandBuffer`(). No idea how that error could be generated there unless the tracking/count is getting messed up by one of the intervening test steps. Any help/insight/feedback would be greatly appreciated!

**Comment by djdeath** [comment]
Looks like master is actually broken for those tests... :( 

Not sure how that happened, I remember testing this. Let me see if I can produce a fix.

**Comment by djdeath** [comment]
Just one error left : 
```
$ ./build/tests/vk_layer_validation_tests --gtest_filter=VkLayerTest.QueryPerformance*
Note: Google Test filter = VkLayerTest.QueryPerformance*
[==========] Running 6 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 6 tests from VkLayerTest
[ RUN      ] VkLayerTest.QueryPerformanceCreation
[       OK ] VkLayerTest.QueryPerformanceCreation (34 ms)
[ RUN      ] VkLayerTest.QueryPerformanceCounterCommandbufferScope
[       OK ] VkLayerTest.QueryPerformanceCounterCommandbufferScope (36 ms)
[ RUN      ] VkLayerTest.QueryPerformanceCounterRenderPassScope
[       OK ] VkLayerTest.QueryPerformanceCounterRenderPassScope (34 ms)
[ RUN      ] VkLayerTest.QueryPerformanceReleaseProfileLockBeforeSubmit
[       OK ] VkLayerTest.QueryPerformanceReleaseProfileLockBeforeSubmit (35 ms)
[ RUN      ] VkLayerTest.QueryPerformanceIncompletePasses
/home/djdeath/src/mesa-src/Vulkan-ValidationLayers/tests/vkrenderframework.cpp:208: Failure
Failed
Expected to succeed but got error: Validation Error: [ VUID-vkGetQueryPoolResults-dataSize-00817 ] Object 0: handle = 0x90000000009, type = VK_OBJECT_TYPE_QUERY_POOL; | MessageID = 0x73375ed4 | vkGetQueryPoolResults() on querypool VkQueryPool 0x90000000009[] specified dataSize 1520 and stride 8, which are incompatible with the specified query type and options. The Vulkan spec states: dataSize must be large enough to contain the result of each query, as described here (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-vkGetQueryPoolResults-dataSize-00817)
[  FAILED  ] VkLayerTest.QueryPerformanceIncompletePasses (50 ms)
[ RUN      ] VkLayerTest.QueryPerformanceResetAndBegin
[       OK ] VkLayerTest.QueryPerformanceResetAndBegin (30 ms)
[----------] 6 tests from VkLayerTest (219 ms total)

[----------] Global test environment tear-down
[==========] 6 tests from 1 test suite ran. (220 ms total)
[  PASSED  ] 5 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] VkLayerTest.QueryPerformanceIncompletePasses

 1 FAILED TEST
```

The change to the `test/vklayertests_others.cpp` in f5b402286b161e260e2a2ecde4e58ccc866a0308 looks odd to me.

**Comment by mark-lunarg** [comment]
@djdeath, bit confused by `VUID-vkGetQueryPoolResults-queryType-03229` which says 
`pData and stride must be multiples of the size of VkPerformanceCounterResultKHR`

I'd interpreted this to mean that since there could be _n_ enabled counters for a specific query, and there were _queryCount_ queries, 

`dataSize = sizeof(VkPerformanceCounterResultKHR) * number of queryresults * queryCount`

and 

`stride = sizeof(VkPerformanceCounterResultKHR) * number of queryresults`

BUT, the example in the spec and the tests in `VkLayerTest.QueryPerformanceIncompletePasses` all use a `stride = sizeof(VkPerofrmanceCounterResultKHR)`, regardless of _queryCount_.

Can you straighten me out, please?

**Comment by djdeath** [comment]
> @djdeath, bit confused by `VUID-vkGetQueryPoolResults-queryType-03229` which says
> `pData and stride must be multiples of the size of VkPerformanceCounterResultKHR`
> 
> I'd interpreted this to mean that since there could be _n_ enabled counters for a specific query, and there were _queryCount_ queries,
> 
> `dataSize = sizeof(VkPerformanceCounterResultKHR) * number of queryresults * queryCount`
> 
> and
> 
> `stride = sizeof(VkPerformanceCounterResultKHR) * number of queryresults`
> 
> BUT, the example in the spec and the tests in `VkLayerTest.QueryPerformanceIncompletePasses` all use a `stride = sizeof(VkPerofrmanceCounterResultKHR)`, regardless of _queryCount_.
> 
> Can you straighten me out, please?

Everything you wrote seems correct to me.

It seems that like other types of queries, you can get less data than available in the queryCount provided to vkGetQueryPoolResults. That VU just seems to reflect the same limitations as the others, which is that the size must be aligned to the element type of the query.

For instance : 
```
VUID-vkGetQueryPoolResults-flags-02827
If VK_QUERY_RESULT_64_BIT is not set in flags, then pData and stride must be multiples of 4
```

Or : 
```
VUID-vkGetQueryPoolResults-flags-00815
If VK_QUERY_RESULT_64_BIT is set in flags then pData and stride must be multiples of 8
```

**Comment by mark-lunarg** [comment]
Right, but I think what's puzzling to me is that it seems the stride has to be _at least_ the size of a single query, which depending on `counterIndexCount`, might be many multiples of 8.  I think this is why 00817 is firing on those tests -- query_size (1520) is greater than the stride (8)?

IOW, shouldn't the stride be the size of an entire single query, and not the size of a `VkPerformanceCounterResultKHR ` (which is what is in those tests)?

**Comment by djdeath** [comment]
> Right, but I think what's puzzling to me is that it seems the stride has to be _at least_ the size of a single query, which depending on `counterIndexCount`, might be many multiples of 8. I think this is why 00817 is firing on those tests -- query_size (1520) is greater than the stride (8)?
> 
> IOW, shouldn't the stride be the size of an entire single query, and not the size of a `VkPerformanceCounterResultKHR ` (which is what is in those tests)?

Oh... I think I see the issue now.

I think you're right, we have an issue with the test. Let me upload a fix.

**Comment by mark-lunarg** [comment]
@djdeath, did you ever get a chance to file a sped issue for the definition of 'stride' for this particular case (performance queries)?

**Comment by djdeath** [comment]
> @djdeath, did you ever get a chance to file a sped issue for the definition of 'stride' for this particular case (performance queries)?

Sorry, forgot about that. I just did.

**Comment by mark-lunarg** [comment]
@djdeath, just rebased this and it's passing CI here.  After the apparent acceptance of your spec change, do you feel that this is ready to go in?   Thanks!

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -7945,6 +7983,12 @@ bool CoreChecks::PreCallValidateCmdCopyQueryPoolResults(VkCommandBuffer commandB
                              "contain VK_QUERY_RESULT_PARTIAL_BIT.",
                              report_data->FormatHandle(queryPool).c_str());
         }
+        if (query_pool_state->createInfo.queryType & VK_QUERY_TYPE_PERFORMANCE_QUERY_INTEL) {
```
Comment (djdeath) [inline layers/core_validation.cpp:7986]: This is probably meant to be `if (query_pool_state->createInfo.queryType == VK_QUERY_TYPE_PERFORMANCE_QUERY_INTEL) {`

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -3755,36 +3764,65 @@ bool CoreChecks::ValidateGetQueryPoolResultsFlags(VkQueryPool queryPool, VkQuery
...
+
+                // These cases intentionally fall through to the default
+                case VK_QUERY_TYPE_ACCELERATION_STRUCTURE_COMPACTED_SIZE_KHR:  // VK_QUERY_TYPE_ACCELERATION_STRUCTURE_COMPACTED_SIZE_NV
+                case VK_QUERY_TYPE_ACCELERATION_STRUCTURE_SERIALIZATION_SIZE_KHR:
+                case VK_QUERY_TYPE_PERFORMANCE_QUERY_INTEL:
+                default:
+                    query_size = 0;
+                    break;
+            }
+            if (query_size && (((query_size > stride)) || ((stride * queryCount) > dataSize))) {
```
Comment (djdeath) [inline layers/core_validation.cpp:3818]: I think there will be a new VU for `query_size > stride`. So maybe let's leave this for somewhere else?

Then I think the correct check for `dataSize` should be : ((queryCount - 1) * stride + query_size) > dataSize

**Comment by djdeath** [comment]
Seems to be all passing, but I found more issues in other tests : https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/2229

File: tests/vklayertests_others.cpp [hunk 4]
```diff
@@ -2933,14 +2936,29 @@ TEST_F(VkLayerTest, QueryPreciseBit) {
...
+        // vk::CmdBeginQuery(m_commandBuffer->handle(), query_pool, 0, VK_QUERY_CONTROL_PRECISE_BIT);
         m_commandBuffer->end();
+
+        const size_t out_data_size = 64;
+        uint8_t data[out_data_size];
+        // The dataSize is too small to return the data
+        m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "VUID-vkGetQueryPoolResults-dataSize-00817");
+        vk::GetQueryPoolResults(m_device->device(), query_pool, 0, 3, 8, &data, 12, 0);
+        m_errorMonitor->VerifyFound();
+        // The stride is too small to return the data
```
Comment (djdeath) [inline tests/vklayertests_others.cpp:2957]: Just need to temporarily disable that test until we get the new VU. Then it's all good :+1:

Comment (mark-lunarg) [reply tests/vklayertests_others.cpp:2952]: :+1

Comment (djdeath) [reply tests/vklayertests_others.cpp]: New VU got released today : VUID-vkGetQueryPoolResults-queryType-04519

**Pull request merged**